pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -107,7 +108,52 @@ func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitG
 	}
 }
 
-func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, _ int) (err error) {
+// If the group key has appeared before, reuse the partial result.
+// If the group key has not appeared before, create empty partial results.
+func (w *HashAggPartialWorker) getPartialResultsOfEachRow(_ *stmtctx.StatementContext, groupKey [][]byte, mapper []AggPartialResultMapper, finalConcurrency int) [][]aggfuncs.PartialResult {
+	numRows := len(groupKey)
+	allMemDelta := int64(0)
+	w.partialResultsBuffer = w.partialResultsBuffer[0:0]
+
+	for i := 0; i < numRows; i++ {
+		finalWorkerIdx := int(murmur3.Sum32(groupKey[i])) % finalConcurrency
+		tmp, ok := mapper[finalWorkerIdx][string(hack.String(groupKey[i]))]
+
+		// This group by key has appeared before, reuse the partial result.
+		if ok {
+			w.partialResultsBuffer = append(w.partialResultsBuffer, tmp)
+			continue
+		}
+
+		// It's the first time that this group by key appeared, create it
+		w.partialResultsBuffer = append(w.partialResultsBuffer, make([]aggfuncs.PartialResult, w.partialResultNumInRow))
+		lastIdx := len(w.partialResultsBuffer) - 1
+		for j, af := range w.aggFuncs {
+			partialResult, memDelta := af.AllocPartialResult()
+			w.partialResultsBuffer[lastIdx][j] = partialResult
+			allMemDelta += memDelta // the memory usage of PartialResult
+		}
+		allMemDelta += int64(w.partialResultNumInRow * 8)
+
+		// Map will expand when count > bucketNum * loadFactor. The memory usage will double.
+		if len(mapper[finalWorkerIdx])+1 > (1<<w.BInMaps[finalWorkerIdx])*hack.LoadFactorNum/hack.LoadFactorDen {
+			w.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice * (1 << w.BInMaps[finalWorkerIdx]))
+			w.BInMaps[finalWorkerIdx]++
+		}
+
+		mapperVal := make([]aggfuncs.PartialResult, w.partialResultNumInRow)
+		for i := range mapperVal {
+			mapperVal[i] = w.partialResultsBuffer[lastIdx][i]
+		}
+		mapper[finalWorkerIdx][string(groupKey[i])] = mapperVal
```

Comment:
Why make a new partialResult slice instead of using w.partialResultsBuffer[lastIdx] directly

Reply:
> Why make a new partialResult slice instead of using w.partialResultsBuffer[lastIdx] directly

Directly assigning w.partialResultsBuffer[lastIdx] to mapper is a shallow copy, if w.partialResultsBuffer[lastIdx] is changed later, mapper will be changed too.